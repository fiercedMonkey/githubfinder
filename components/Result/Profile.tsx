import {
  Avatar,
  Box,
  Button,
  Card,
  CardContent,
  Typography,
} from "@mui/material";
import React, { FunctionComponent } from "react";
import { userRepoContextType } from "@/types/Repo";
import Link from "next/link";

type ProfileProps = {
  user: userRepoContextType["userData"];
};

export const Profile: FunctionComponent<ProfileProps> = ({ user }) => {
  const foo = 1;

  return (
    <>
      {user && (
        <Card sx={{ marginBottom: "2rem" }}>
          <CardContent>
            <Box component={"div"} sx={{ display: "flex" }}>
              <Box component={"div"} sx={{ margin: "1rem" }}>
                <Avatar
                  alt={user.login}
                  src={user.avatar_url}
                  sx={{ width: 56, height: 56 }}
                />
              </Box>
              <Box component={"div"} sx={{ margin: "1rem" }}>
                <Typography variant="h5" component="div">
                  {user.login}
                </Typography>
                <Typography sx={{ mb: 1.5 }} color="text.secondary">
                  Repositories: {user.public_repos}
                </Typography>
                <Link href={user.html_url} legacyBehavior>
                  <a target={"_blank"}>
                    <Button size="small">Visit Profile on Github</Button>
                  </a>
                </Link>
              </Box>
            </Box>
          </CardContent>
        </Card>
      )}
    </>
  );
};
