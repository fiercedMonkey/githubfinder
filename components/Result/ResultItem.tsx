import React, { FunctionComponent } from "react";
import { RepoData } from "@/types/Repo";
import { TableCell, TableRow } from "@mui/material";
import { useRouter } from "next/router";

type ResultItemProps = {
  repoData: RepoData;
  align?: "right";
};

export const ResultItem: FunctionComponent<ResultItemProps> = ({
  repoData,
  align,
}) => {
  const onTableRowClick = () => {
    window.open(repoData.html_url, "_blank");
  };

  return (
    <>
      <TableRow
        hover
        tabIndex={-1}
        key={repoData.name}
        onClick={onTableRowClick}
        sx={{ cursor: "pointer" }}
      >
        <TableCell align={align}>{repoData.name}</TableCell>
        <TableCell align={align}>{repoData.stargazers_count}</TableCell>
        <TableCell align={align}>{repoData.forks_count}</TableCell>
        <TableCell align={align}>{repoData.html_url}</TableCell>
      </TableRow>
    </>
  );
};
