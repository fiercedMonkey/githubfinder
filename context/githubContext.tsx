import { createContext, useContext, useState } from "react";
import axios from "axios";
import { userRepoContextType } from "@/types/Repo";
import { useRouter } from "next/router";

const githubContextInitialValue = {
  notFound: false,
  repoData: null,
  userData: null,
  getUserRepos: () => {},
  clearState: () => {},
};

const GithubContext = createContext<userRepoContextType>(
  githubContextInitialValue
);

export function useGithub() {
  return useContext(GithubContext);
}

export function GithubProvider({ children }: any) {
  const router = useRouter();
  const [notFound, setNotFound] = useState(false);
  const [repoData, setRepoData] =
    useState<userRepoContextType["repoData"]>(null);
  const [userData, setUserData] =
    useState<userRepoContextType["userData"]>(null);

  const clearState = () => {
    setUserData(null);
    setRepoData(null);
  };
  const getUserRepos = async (username: string) => {
    await axios
      .get(`https://api.github.com/users/${username}`)
      .then((res) => {
        setUserData(res.data);
        setNotFound(res.status !== 200);
      })
      .catch((e) => setNotFound(e.response.status === 404));

    await axios
      .get(`https://api.github.com/users/${username}/repos`, {
        params: { per_page: 100 },
      })
      .then((res) => {
        setRepoData(res.data);
        router.push("/result");
      })
      .catch((e) => e);
  };

  const value: userRepoContextType = {
    notFound,
    repoData,
    userData,
    getUserRepos,
    clearState,
  };

  return (
    <>
      <GithubContext.Provider value={value}>{children}</GithubContext.Provider>
    </>
  );
}
