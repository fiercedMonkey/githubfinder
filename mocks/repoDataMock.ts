import { userRepoContextType } from "@/types/Repo";

export const repoDataMock: userRepoContextType["repoData"] = [
  {
    name: "action-check-pr-title",
    html_url: "https://github.com/fiercedMonkey/action-check-pr-title",
    stargazers_count: 0,
    forks_count: 0,
  },
  {
    name: "epiloit-githubFinder",
    html_url: "https://github.com/fiercedMonkey/epiloit-githubFinder",
    stargazers_count: 0,
    forks_count: 0,
  },
];
