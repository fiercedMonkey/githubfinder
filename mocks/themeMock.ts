import { createTheme, Theme } from "@mui/material/styles";

export const themeMock: Theme = createTheme({
  palette: {
    mode: "dark",
  },
});
