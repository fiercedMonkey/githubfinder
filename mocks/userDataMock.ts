import { userRepoContextType } from "@/types/Repo";

export const userDataMock: userRepoContextType["userData"] = {
  login: "fiercedMonkey",
  avatar_url: "https://avatars.githubusercontent.com/u/15922918?v=4",
  html_url: "https://github.com/fiercedMonkey",
  public_repos: 2,
};
