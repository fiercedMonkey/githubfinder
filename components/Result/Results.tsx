import React, { FunctionComponent } from "react";
import { ResultItem } from "@/components/Result/ResultItem";
import { userRepoContextType } from "@/types/Repo";
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";

type ResultsProps = {
  repoData: userRepoContextType["repoData"];
};

interface Column {
  id: string;
  label: string;
  minWidth?: number;
  align?: "right";
}

const columns: readonly Column[] = [
  { id: "name", label: "Name" },
  { id: "stars", label: "Stars" },
  { id: "forks", label: "Forks" },
  { id: "link", label: "Link" },
];

export const Results: FunctionComponent<ResultsProps> = ({ repoData }) => {
  return (
    <Paper sx={{ width: "900px", overflow: "hidden" }}>
      <TableContainer>
        <Table
          stickyHeader
          aria-label="sticky table"
          sx={{ border: "1px solid grey" }}
        >
          <TableHead>
            <TableRow>
              {columns.map((column) => (
                <TableCell
                  key={column.id}
                  align={column.align}
                  style={{ minWidth: column.minWidth }}
                >
                  {column.label}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {repoData &&
              repoData.map((repo) => (
                <ResultItem repoData={repo} key={repo.name} />
              ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Paper>
  );
};
