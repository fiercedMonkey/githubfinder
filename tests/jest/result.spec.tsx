import { render } from "@testing-library/react";
import Result from "../../pages/result";

import React from "react";
import { createTheme, ThemeProvider } from "@mui/material/styles";
import { CssBaseline } from "@mui/material";
import Home from "@/pages";
import { GithubProvider } from "@/context/githubContext";
import { repoDataMock } from "@/mocks/repoDataMock";
import { useRouter } from "next/router";
import { userDataMock } from "@/mocks/userDataMock";

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const darkTheme = createTheme({
  palette: {
    mode: "dark",
  },
});

describe("Home", () => {
  it("renders resultpage", () => {
    const mockRouter = {
      replace: jest.fn(),
    };

    (useRouter as jest.Mock).mockReturnValue(mockRouter);

    const { container } = render(
      <ThemeProvider theme={darkTheme}>
        <CssBaseline />
        <GithubProvider value={[repoDataMock, userDataMock]}>
          <Result />
        </GithubProvider>
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
