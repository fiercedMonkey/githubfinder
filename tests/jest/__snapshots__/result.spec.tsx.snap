// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders resultpage 1`] = `
<div>
  <div
    class="MuiBox-root css-gmuovu"
  >
    <div
      class="MuiBox-root css-1l4w6pd"
    >
      <div
        class="MuiBox-root css-6n2mi5"
      >
        <a
          href="/"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-4wjvfs-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Back to search
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
      <div
        class="MuiBox-root css-i51og3"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-7x6vwb-MuiPaper-root"
        >
          <div
            class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
          >
            <table
              aria-label="sticky table"
              class="MuiTable-root MuiTable-stickyHeader css-1ut45vx-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1pgphf0-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-1gw840t-MuiTableCell-root"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-1gw840t-MuiTableCell-root"
                    scope="col"
                  >
                    Stars
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-1gw840t-MuiTableCell-root"
                    scope="col"
                  >
                    Forks
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-1gw840t-MuiTableCell-root"
                    scope="col"
                  >
                    Link
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              />
            </table>
          </div>
        </div>
        <div
          class="MuiTablePagination-root css-n2sbe9-MuiTablePagination-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
          >
            <div
              class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
            />
            <p
              class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
            >
              0â€“0 of 0
            </p>
            <div
              class="MuiTablePagination-actions"
            >
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-b57xgn-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to previous page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-b57xgn-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                tabindex="-1"
                title="Go to next page"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
